### PR TITLE
React to Options change

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
Reacting to aspnet/Options#249, which ensures that more services will be in the collection.

### CODE REVIEW

File: test/Microsoft.AspNetCore.Mvc.Localization.Test/Internal/MvcLocalizationServiceCollectionExtensionsTest.cs [hunk 1]
```diff
@@ -31,6 +31,37 @@ public void AddLocalizationServices_AddsNeededServices()
 
             // Assert
             Assert.Collection(collection,
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Localization.Test/Internal/MvcLocalizationServiceCollectionExtensionsTest.cs]: Could we change these tests to use the pattern from https://github.com/aspnet/Mvc/blob/dev/test/Microsoft.AspNetCore.Mvc.Core.Test/DependencyInjection/MvcCoreServiceCollectionExtensionsTest.cs? We really don't want to verify what services dependencies add, just the ones we care about.

**Comment by ryanbrandenburg** [comment]
:up::date:

File: test/Microsoft.AspNetCore.Mvc.Localization.Test/Internal/MvcLocalizationServiceCollectionExtensionsTest.cs [hunk 2]
```diff
@@ -17,6 +18,15 @@ namespace Microsoft.AspNetCore.Mvc.Localization.Internal
 {
     public class MvcLocalizationServicesTest
     {
+        private readonly static HtmlTestEncoder HtmlEncoderInstance = new HtmlTestEncoder(); 
+
+        private readonly IDictionary<Type, Tuple<object, ServiceLifetime>> ExpectedServices = new Dictionary<Type, Tuple<object, ServiceLifetime>> {
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Localization.Test/Internal/MvcLocalizationServiceCollectionExtensionsTest.cs]: We really don't care about service lifetimes. Can we stop asserting these?

File: test/Microsoft.AspNetCore.Mvc.Localization.Test/Internal/MvcLocalizationServiceCollectionExtensionsTest.cs [hunk 3]
```diff
@@ -17,6 +18,15 @@ namespace Microsoft.AspNetCore.Mvc.Localization.Internal
...
+            { typeof(IHtmlLocalizerFactory), new Tuple<object, ServiceLifetime>(typeof(HtmlLocalizerFactory), ServiceLifetime.Singleton)},
+            { typeof(IHtmlLocalizer<>),  new Tuple<object, ServiceLifetime>(typeof(HtmlLocalizer<>), ServiceLifetime.Transient)},
+            { typeof(IViewLocalizer), new Tuple<object, ServiceLifetime>(typeof(ViewLocalizer), ServiceLifetime.Transient) },
+            { typeof(IConfigureOptions<RazorViewEngineOptions>), new Tuple<object, ServiceLifetime>(null, ServiceLifetime.Singleton) }
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Localization.Test/Internal/MvcLocalizationServiceCollectionExtensionsTest.cs]: Skip this one - this is not useful.

File: test/Microsoft.AspNetCore.Mvc.Localization.Test/Internal/MvcLocalizationServiceCollectionExtensionsTest.cs [hunk 4]
```diff
@@ -105,73 +61,22 @@ public void AddCustomLocalizers_BeforeAddLocalizationServices_AddsNeededServices
                 LanguageViewLocationExpanderFormat.Suffix,
                 setupAction: null);
 
+            var expected = ExpectedServices;
+            expected[typeof(IHtmlLocalizerFactory)] = new Tuple<object, ServiceLifetime>(typeof(TestHtmlLocalizerFactory), ServiceLifetime.Singleton);
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Localization.Test/Internal/MvcLocalizationServiceCollectionExtensionsTest.cs]: Can we just make a new dictionary instead?

File: test/Microsoft.AspNetCore.Mvc.Localization.Test/Internal/MvcLocalizationServiceCollectionExtensionsTest.cs [hunk 5]
```diff
@@ -105,73 +61,22 @@ public void AddCustomLocalizers_BeforeAddLocalizationServices_AddsNeededServices
...
-                service =>
-                {
-                    Assert.Equal(typeof(IConfigureOptions<RazorViewEngineOptions>), service.ServiceType);
-                    Assert.Equal(ServiceLifetime.Singleton, service.Lifetime);
-                });
+            AssertContainsExpected(collection, expected);
         }
 
         [Fact]
         public void AddCustomLocalizers_AfterAddLocalizationServices_AddsNeededServices()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Localization.Test/Internal/MvcLocalizationServiceCollectionExtensionsTest.cs]: Let's delete this test. It's not testing any sort of useful behavior (in reality, services that get added after will get resolved by the DI container).

File: test/Microsoft.AspNetCore.Mvc.Localization.Test/Internal/MvcLocalizationServiceCollectionExtensionsTest.cs [hunk 6]
```diff
@@ -283,61 +116,102 @@ public void AddLocalizationServicesWithLocalizationOptions_AddsNeededServices()
...
+                    AssertContainsSingleType(services, expectedService.Key, implementationType, expectedService.Value.Item2);
+                }
+                else if(expectedCount != null)
                 {
-                    Assert.Equal(typeof(IHtmlLocalizer<>), service.ServiceType);
-                    Assert.Equal(typeof(HtmlLocalizer<>), service.ImplementationType);
-                    Assert.Equal(ServiceLifetime.Transient, service.Lifetime);
-                },
-                service =>
+                    AssertContainsMultiple(services, expectedService.Key, expectedCount.Value, expectedService.Value.Item2);
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Localization.Test/Internal/MvcLocalizationServiceCollectionExtensionsTest.cs]: We don't have any multi registrations \ instances here. Can we just remove these for now?

File: test/Microsoft.AspNetCore.Mvc.Localization.Test/Internal/MvcLocalizationServiceCollectionExtensionsTest.cs [hunk 7]
```diff
@@ -283,61 +116,102 @@ public void AddLocalizationServicesWithLocalizationOptions_AddsNeededServices()
...
-                    Assert.Equal(typeof(IConfigureOptions<RazorViewEngineOptions>), service.ServiceType);
-                    Assert.Equal(ServiceLifetime.Singleton, service.Lifetime);
-                },
-                service =>
+        private static void AssertContainsExpected(IServiceCollection services, IDictionary<Type, Tuple<object, ServiceLifetime>> expectedServices)
+        {
+            foreach(var expectedService in expectedServices)
+            {
+                var implementationType = expectedService.Value.Item1 as Type;
+                var expectedCount = expectedService.Value.Item1 as int?;
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Localization.Test/Internal/MvcLocalizationServiceCollectionExtensionsTest.cs]: We don't need this. It's basically `Dictionary<Type, Type>`.

**Pull request closed without merging**